File: algorithm-js/sorting/quick-sort.js
var arr = [5, 2, 4, 6, 1, 3];

console.log(arr);
console.log(quicksort(arr));

function quicksort(array) {
    var less = [], greater = [];
    var pivot;
    if(array.length <= 1) {
        return array;
    } else {
        pivot = array[0];
        var l = array.length;
        for(var i = 1; i < l; ++i) {
            if(array[i] < pivot) {
                less.push(array[i]);
            } else {
                greater.push(array[i]);
            }
        }
        var sorted = [];
        return sorted.concat(quicksort(less), pivot, quicksort(greater));
    }
}

File: algorithm-c/sorting/quicksort/quicksort.c
#include <stdio.h>

inline void swap(int* a, int* b)
{
    int tmp;
    tmp = *a; *a = *b; *b = tmp;
}

int main()
{
    void pnt(int arr[], int length);
    void quicksort(int* arr, int left, int right);

    int array[] = {8, 4, 7, 5, 1, 0, 8, 9, 2, 6, 9, 5, 3};
    pnt(array, 13);
    quicksort(array, 0, 12);
    pnt(array, 13);
    return 0;
}

void pnt(int arr[], int length)
{
    int i = 0;
    while(i < length)
    {
        printf("%d%s", arr[i], " ");
        ++i;
    }
    printf("\n");
}

void quicksort(int* arr, int left, int right)
{
    int pivot = *(arr + left);
    swap(arr + left, arr + right);
    int storeIndex = left;
    int i = 0;
    for(i = left; i <= right - 1; ++i)
    {
        if(arr[i] < pivot)
        {
            swap(arr + storeIndex, arr + i);
            storeIndex++;
        }
    }
    swap(arr + storeIndex, arr + right);
    if((storeIndex - left) > 1) quicksort(arr, left, storeIndex);
    if((right - storeIndex - 1) > 1) quicksort(arr, storeIndex + 1, right);
}
